### DISCUSSION TITLE
General question about source gen

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [IEvangelist/blazorators#26] [Q&A] [answered]
No discussions area, but this is more of a question really than any issue. Watching the [YouTube](https://youtube.com/watch?v=U8XriF2zxS0) from earlier and it is very interesting to me. This is somewhat along the lines of what I think of for code generation.

Some common task, goal, boilerplate, which is going to repeat, so add the hooks that allow for CG, right. I've got a case where I define some _CSharp_ first class `Enumeration` types, very much analog to _Java_ `Enumeration` types, and sometimes I enable them to be functionally Bitwise; that's my boilerplate code that I like to CG. So I afford some base classes including an Attribute that I can key on during compilation and spit out the corresponding CG'ed code, right.

Some other examples as of late along the lines of CG'ed regular expressions, I'm not even sure what use case that addresses, TBH.

However, in this case what is the code upon which you trigger, and does it really vary that much from build to build? Contingent upon a js dependency, let's say, so, ostensibly, yes, it may vary sometimes. Can you simply not do some 'reflection' after a sort on the js itself to generate code?

Also, why what is effectively a fancier [_T4 text replacement_](https://en.wikipedia.org/wiki/Text_Template_Transformation_Toolkit) and not something like a [Roslyn](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/), where you _CSharp_-up the language level, or at least _Intermeriate Language_ (IL) level, primitives around the same?

If I understand the thing correctly. Thanks and great presentation earlier!

**Comment by IEvangelist** [comment, marked as answer]
Hi @mwpowellhtx,

The source that is generated is currently dependent on this TypeScript type declaration:

- https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts

If you open that file and do a search for `interface Storage {`, you'll find the declaration for it. It's different that T4 as it's not based on templates, and it uses Roslyn. Does that make sense?

**Comment by mwpowellhtx** [comment]
Ah well I must say, about the [25:26](https://youtube.com/watch?v=U8XriF2zxS0&t=25m26s) mark is what made me think of this being more of a _boilerplate T4 text replacement_ approach than what I would think of as a true, using the _Roslyn CSharp_ primitives to assemble your source approach.

For comparison, have a gander at my [BitwiseOperatorOverloadsPartialGenerator](https://github.com/mwpowellhtx/Ellumination.Collections/blob/master/src/Ellumination.Collections.Enumerations.Generators/BitwiseOperatorOverloadsPartialGenerator.cs), curious to hear your thoughts on this sort of _Roslyn_ approach. For which I also sometimes refer to the [Roslyn Quoter](http://roslynquoter.azurewebsites.net/) quite frequently when I am deep diving in that regard, and pare things back a bit, taylor, factor, etc; but as a starting point, the _Quoter_ is quite fantastic as a reference.

Obviously with the _Blazorators_, there are a number of moving parts involved, if you will, considering the _TypeScript_ (i.e. _js_), _Blazor_, _CSharp_, and moving parts in between.

Cheers, man, and if we might collaborate, maybe there's a partnership in it for us in the future, who knows.

**Reply by IEvangelist** [reply]
> Ah well I must say, about the [25:26](https://youtube.com/watch?v=U8XriF2zxS0&t=25m26s) mark is what made me think of this being more of a boilerplate T4 text replacement approach than what I would think of as a true, using the Roslyn CSharp primitives to assemble your source approach.

That was just for the `JSAutoInteropAttribute` and `JSAutoGenericInteropAttribute`. There are only a few places where strings are used like that for adding source.
